man/alfa.1
.\" Manpage for alfa.
.TH alfa 1 "20 Aug 2025" "2.3" "alfa man page"
.SH NAME
\fBalfa\fR \- automated line fitting algorithm
.SH SYNOPSIS
\fBalfa\fR [\fIOPTION\fR] [\fIVALUE\fR]... [\fIFILE\fR]
.SH DESCRIPTION
\fBalfa\fR rapidly fits emission line spectra, using a genetic algorithm to optimise fitting parameters.  It is intended to be entirely automated, but while the default values should work well in many situations, a good fit to your observed spectrum may require some adjustments to the input parameters.  It is optimised for optical spectra, but can be applied to any wavelength range if a suitable line catalogue is provided.
.PP
\fBalfa\fR reads one dimensional spectra in either plain text or FITS format.  Plain text input should consist of two columns, giving wavelength and flux.  It can also read data cubes and row-stacked spectra in FITS format.  Results are written out in plain text, to files containing the fit (total fit, continuum-subtracted original, continuum, sky lines and residuals), and the line flux measurements.
.SH OPTIONS

.TP
.B \-b, \-\-bad-data [REAL]
If all values in a spectrum are below the value specified, \fBalfa\fR will not fit it.  Most useful for avoiding wasting time on low signal regions of data cubes.

.TP
.B \-\-citation
Prints out the bibliographic details of the paper to cite if you use \fBalfa\fR in your research.

.TP
.B \-cl, \-\-clobber
\fBalfa\fR's default behaviour is that it will not overwrite any pre-existing output file. If this option is specified, then the output file will be overwritten.

.TP
.B \-\-collapse
Sums all spectra in multi-dimensional FITS files into a single spectrum.  Spectra whose peak value is below any value specified with the \fB--bad-data\fR option are excluded.  \fB--collapse\fR has no effect on 1D data.

.TP
.B \-cs, \-\-chunk\-size [INTEGER]
Sets the size in pixels of the chunks into which the spectrum is split for fitting. Default: 400

.TP
.B \-cw, \-\-continuum\-window [INTEGER]
ALFA estimates the continuum flux at each wavelength as the median of flux values within 3 standard deviations of the mean flux in a window centred on the wavelength. This option can be used to set the size of the window. The value must be an odd integer. Default: 101

.TP
.B \-dl, \-\-detection-limit [REAL]
Sets the factor by which a line's flux must exceed the locally estimated noise level for it to be considered a detection.  Default: 3.0

.TP
.B \-el, \-\-exclude-line [REAL]
When reading in the line catalogues, any wavelengths indicated with this option will be ignored.  For example, if H alpha were saturated, it could be excluded from the fit with --exclude-line 6562.77.  Any number of lines can be excluded by repeating the option with the appropriate wavelengths.

.TP
.B \-fc, \-\-flux-column [INTEGER]
When reading FITS tables, this option can be used to specify in which table column the flux values are. For any other file type it will have no effect. Default: 2

.TP
.B \-g, \-\-generations [INTEGER]
The number of generations used in the genetic algorithm. Default: 500

.TP
.B \-n, \-\-normalise [VALUE]
Normalise to Hb=100 assuming that F(Hb)=\fIVALUE\fR.  If \fIVALUE\fR is zero, no normalisation is applied.  If this option is not specified, fluxes are normalised using the measured value of Hb if it is detected, and not normalised otherwise.

.TP
.B \-nc, \-\-no\-continuum
In case the spectrum you are fitting is already continuum subtracted, setting this option skips \fBalfa\fR's continuum fitting.

.TP
.B \-o, \-\-output\-dir [DIRECTORY]
The directory in which to put the output files. Default: current working directory.

.TP
.B \-of, \-\-output\-format [VALUE]
The format of the output files. Valid values are text, csv, fits, latex. If you are using \fBalfa\fR in conjunction with \fBneat\fR, plain text and fits formats can currently be read in. Default: fits

.TP
.B \-pr, \-\-pressure [REAL]
The fraction of the population retained from each generation. The product of the pressure and the population size should be an integer.  Default: 0.3

.TP
.B \-ps, \-\-populationsize [INTEGER]
The size of the population used in the genetic algorithm. Default: 30

.TP
.B \-rb, \-\-rebin [INTEGER]
Rebin the input spectrum by the specified factor. Useful for high resolution spectra where line profiles are not instrumental but kinematic. This option is currently only implemented for 1d spectra.

.TP
.B \-rg, \-\-resolution\-guess [VALUE]
Initial guess for the resolution [lambda/delta lambda]. Default: estimated using the sampling of the input spectrum, assuming that it is Nyquist sampled.

.TP
.B \-rtol1, \-\-resolution\-tolerance\-1 [VALUE]
Variation allowed in resolution in first pass. Default: equal to 0.9 x resolution guess.

.TP
.B \-rtol2, \-\-resolution\-tolerance\-2 [VALUE]
Variation allowed in resolution in second pass. Default: 500.

.TP
.B \-skyc, --sky-catalogue; \-sc, --strong-catalogue; \-dc, --deep-catalogue [FILENAME]
The files containing the line catalogues to be used for the removal of sky lines, the estimation of velocity and resolution, and the full line fitting.  The default catalogues are stored in /usr/share/alfa .  If you wish to create your own catalogue, the required format is that each line should be 85 characters wide, with a wavelength in the first column, and the rest of the characters are not used by the code but are transferred to the output files.  They can thus be used, as in the supplied catalogues, for line transition data.  To use the default catalogues but exclude some lines, the \fB--exclude-line\fR option can be used.

.TP
.B \-ss, \-\-subtract\-sky
Fit and subtract night sky emission lines before fitting nebular emission lines.

.TP
.B \-ul, \-\-upper-limits
Write out upper limits for all lines searched for and not detected. The upper limit is calculated as the detection limit (default 3, set by \-\-detection\-limit) multiplied by the uncertainty. In FITS output, this option is ignored - all upper limits are written out, and are indicated by negative flux values. If the FITS file is later analysed with \fBneat\fR, these negative values are simply ignored.

.TP
.B \-vg, \-\-velocity\-guess [VALUE]
Initial guess for the velocity of the object [km/s]. Default: 0.

.TP
.B \-vtol1, \-\-velocity\-tolerance\-1 [VALUE]
Variation allowed in velocity in first pass of the fitting. Default: 900km/s

.TP
.B \-vtol2, \-\-velocity\-tolerance\-2 [VALUE]
Variation allowed in velocity in second pass of the fitting. Default: 60km/s

.TP
.B \-wc, \-\-wavelength-column [INTEGER]
When reading FITS tables, this option can be used to specify in which table column the wavelength values are. For any other file type it will have no effect. Default: 1

.TP
.B \-ws, \-\-wavelength\-scaling [VALUE]
\fBalfa\fR checks the units of FITS file headers to set the wavelength scale, defaulting to assuming Angstroms if no keyword is present.  If your input spectrum is not in Angstroms, use this option to set the value by which wavelengths should be multiplied to convert them to A.  For example, \-ws 10.0 would apply if your spectra have wavelengths in nm.

.SH ALGORITHM

\fBalfa\fR works in three stages.  First, it estimates and subtracts the continuum.  Second, it estimates the resolution of the spectra and the velocity of the object.  And third, it fits the emission lines.  These stages work as follows:

.TP
.B Continuum subtraction

\fBalfa\fR estimates the continuum using a percentile filter, taking the 25th percentile in a moving window 101 pixels wide.  Currently these values are hard-coded, but will be user-configurable in a future release.  Regions such as the Balmer and Paschen jumps may be poorly fitted by this method if the spectral resolution is low and the continuum gradient is changing fast.  Broad stellar emission lines and telluric absorption features may also not be well fitted by this method.  Inspection of the fitted continuum is recommended.

.TP
.B Estimation of the resolution and velocity

If no relevant command line options are specified, \fBalfa\fR begins by assuming that the velocity of the object is zero, and that the spectrum is Nyquist-sampled.  It then carries out a fit on a subset of strong lines, using the genetic algorithm described below, to obtain an overall estimate for the velocity and the resolution.  If necessary, the initial guesses can be specified with the \fB-vg\fR and \fB-rg\fR options described above, and the parameter space for the fine tuning can be specified with \fB-vtol1\fR and \fB-rtol1\fR.

.TP
.B Fitting of the emission lines

With the continuum subtracted and the resolution and velocity estimated, \fBalfa\fR divides the spectrum up into chunks 440 pixels wide, with 20 pixels at either end overlapping with adjacent chunks.  Then the genetic algorithm fits all lines from the deep catalogue that fall within the central 400 pixels, with the overlap regions providing the full line profile for lines close to the edge of the chunk.  The initial guess for the resolution and velocity are taken from the global estimate for the first chunk, and from the preceding chunk's fine tuned value for all succeeding chunks.

With the parameters optimised in each chunk, uncertainties are estimated using the root mean square of the residuals in a 20 pixel window, exlucing the two largest residuals to mitigate against overestimated uncertainties in the neighbourhood of bad pixels or strong lines.

.SH INPUT FILES
\fBalfa\fR can read either plain text files or FITS format files.  For plain text, the file should contain a wavelength and a flux, with the wavelength in the same units as the line catalogues (the default catalogues have wavelengths in Angstroms).  FITS files are read using the CFITSIO library, so any FITS-compliant file should be fine.  However, a surprisingly large fraction of all FITS files do not comply with the standard, so in case of problems, trying using \fBfitsverify\fR to check your FITS file.

The FITS file can have one, two or three dimensions.  If it has two, it is assumed to be in Row-Stacked Spectra (RSS) format, while if it has three, it is assumed to be a data cube with two axes representing spatial dimensions and the third representing the spectral dimension.

If you don't want to fit the whole dataset, you can specify the range of pixels on each axis that you want \fBalfa\fR to read in by appending the filename with [\fIstartx:endx,starty:endy\fR].  This functionality is part of the CFITSIO library, and the format is described at https://heasarc.gsfc.nasa.gov/docs/software/fitsio/c/c_user/node97.html.  \fBalfa\fR itself does not read in the coordinates of the section, and so the output file numbering starts from 1 on each axis regardless of where the image section actually started.  Improved support for image sections is planned for future releases.

.SH OUTPUT FILES
\fBalfa\fR can produce output in plain text, csv, latex or FITS format. FITS is the default and recommended format; in future releases, output formats other than FITS will be deprecated and eventually removed.

When producing plain text, csv or latex output, \fBalfa\fR will produce two files, a fit file and a line list file. When producing FITS format, a single file is created, with three extensions, for the fit, the linelist, and a third extension containing the Hbeta flux, number of lines detected, the radial velocity and average resolution over the spectrum. Output filenames are based on the input filename, with the extensions noted below.

.TP
.B The fit file (_fit.txt, _fit.csv, _fit.tex) or extension (.fits):

The fit file contains the best fitting synthesised spectrum.  It contains seven columns, representing the wavelength, the input spectrum, the fitted spectrum, the original after continuum subtraction, the estimated continuum, the fluxes of sky lines, and the residuals.  Thus, to see the fitted spectrum, you need to plot columns 1 and 3 of this file.  In gnuplot, one can compare the input and fitted spectra using this command:
    plot 'filename_fit' w l, '' using 1:3 w l

.TP
.B The lines file (_lines.txt, _lines.csv, _lines.tex) or extension (.fits):

This file contains four columns with parameters of the fitted lines - the observed wavelength, the rest wavelength, the flux, and the uncertainty estimated from the residuals.  This file can be read directly by \fBneat\fR, which determines abundances for photoionised nebulae. If latex format is requested, the file additionally contains atomic data details not currently available in the other formats.

Lines are flagged as blended if they are separated by less than 0.75x the estimated resolution. For blended lines, the total flux is given with the first line of the feature. For the rest of the blended lines, the flux is given as "*" in text output, and 0 in fits output, or as -999 in the case that the blend is flagged as a non-detection.

For RSS files and data cubes, \fBalfa\fR currently produces output for each pixel. Thus, for a large data cube you may end up with tens of thousands of files in the output directory. A routine to combine these outputs into image maps will be provided in a future release of \fBalfa\fR.

.SH USAGE NOTES

\fBalfa\fR's default parameters are supposed to work in most cases, but sometimes you might find that it does not converge on the correct wavelength solution.  It searches initially for velocities in the range +/-900km/s, which is very large for Galactic objects.  So, running the code with \fB--resolution-tolerance-1 100.\fR or so may improve your results.

The genetic parameters (population size, number of generations, pressure) are likely to be suitable for most cases.  There is no algorithm yet known for optimising these parameters in a genetic algorithm, so changing them requires trial and error.  In spectra of regions with lots of emission lines, such as 4000-4500 Angstrom, increasing the number of generations can result in a better fit.

.SS Error codes

If \fBalfa\fR encounters an error it will return one of the following status codes:

 100 incomplete or invalid command line option
 101 input file does not exist
 102 output directory does not exist
 103 error opening FITS file
 104 no axes found in input file
 105 too many axes found in input file
 106 error reading FITS keywords
 107 error writing FITS output
 108 no line catalogue file specified
 109 line catalogue not found

 200 all fluxes below baddata limit
 201 no known emission lines in wavelength range
 202 failed to estimate velocity and resolution - no lines found

.SH SEE ALSO
neat
.SH BUGS
If reporting a bug, please state which version of \fBalfa\fR you were using, and include input and any output files produced if possible.
.SH AUTHOR
Roger Wesson


source/alfa.f90
program alfa
!ALFA, the Automated Line Fitting Algorithm
!Copyright (C) 2013- Roger Wesson

!This program is free software: you can redistribute it and/or modify
!it under the terms of the GNU General Public License as published by
!the Free Software Foundation, either version 3 of the License, or
!(at your option) any later version.

!This program is distributed in the hope that it will be useful,
!but WITHOUT ANY WARRANTY; without even the implied warranty of
!MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
!GNU General Public License for more details.

!You should have received a copy of the GNU General Public License
!along with this program.  If not, see <http://www.gnu.org/licenses/>.

use mod_readfiles
use mod_functions
use mod_types
use mod_continuum
use mod_fit
use mod_uncertainties
use mod_commandline
use mod_globals
use mod_output

! openmp variables
implicit none
integer :: omp_get_thread_num, omp_get_num_threads

! variables written to within parallel block cannot be global

character(len=512) :: outputbasename
type(linelist), dimension(:), allocatable :: fittedlines, fittedlines_section, skylines, skylines_section
type(spectrum), dimension(:), allocatable :: realspec, fittedspectrum, spectrumchunk, skyspectrum, continuum, stronglines, originalcopy
real :: redshiftguess, redshiftguess_initial, redshiftguess_overall ! redshiftguess_initial is the user-specified value, used in the initial fit which determines redshiftguess_overall.  that is then used in the chunks to find redshift
real :: resolutionguess, resolutionguess_initial ! resolutionguess_initial is the user-specified value, used in the initial fit to determine resolutionguess.
integer :: linearraypos,totallines, startpos, endpos, nlines
real :: normalisation, hbetaflux
integer :: threadnumber

c=299792.458 !km/s
!default values in absence of user specificed guess
redshiftguess=0.0 !km/s
redshiftguess_initial=0.0
resolutionguess=0.0 !lambda/deltalambda, determined assuming nyquist sampling if not specified
resolutionguess_initial=0.0
rtol1=0.d0 !variation allowed in resolution on first pass.  determined later, either from user input, or to be equal to resolution guess.
rtol2=500. !second pass
vtol1=0.0006 !variation allowed in velocity (expressed as redshift) on first pass. 0.0006 = 180 km/s
vtol2=0.0002 !second pass. 0.0002 = 60 km/s
baddata=0.d0
bdcount=0
wavelengthscaling=1.d0
detectionlimit=3.0
rebinfactor=1
continuumwindow=101
chunksize=400

tablewavelengthcolumn=1
tablefluxcolumn=2

outputformat="fits"

stronglinelistfile=""
deeplinelistfile=""
skylinelistfile=""

outputdirectory="./"
imagesection=""

collapse=.false.
messages=.false.
clobber=.false.

popsize=30
pressure=0.3 !pressure * popsize needs to be an integer
generations=500

! start

print *,"ALFA, the Automated Line Fitting Algorithm"
print *,"version ",VERSION

print *
print *,gettime(),"starting code"

! random seed

call initialize()

! read command line. no variables to pass except the two which are not global

call readcommandline(redshiftguess_initial,resolutionguess_initial,normalisation)

! convert from velocity to redshift

redshiftguess_initial=1.+(redshiftguess_initial/c)

! read in spectrum to fit and line list

print *
print *,gettime(),"reading in file ",trim(spectrumfile),":"

!call subroutine to read in the data.  input is filename, output is 3D array containing data, length of dimensions dependent on whether file was 1D, 2D or 3D.

call readdata()

minimumwavelength = wavelengths(1)
maximumwavelength = wavelengths(size(wavelengths))
spectrumlength = size(wavelengths)

! set catalogue if not already specified
! select optical catalogue if min wavelength greater than 1000, IR if max wavelength less than 1000

if (minimumwavelength.gt.1000) then
  if (stronglinelistfile.eq."") then
    stronglinelistfile=trim(PREFIX)//"/share/alfa/optical_strong.cat"
    print *,gettime(),"optical strong line catalogue selected"
  endif
  if (deeplinelistfile.eq."") then
    deeplinelistfile=trim(PREFIX)//"/share/alfa/optical_deep.cat"
    print *,gettime(),"optical deep line catalogue selected"
  endif
else
  if (stronglinelistfile.eq."") then
    stronglinelistfile=trim(PREFIX)//"/share/alfa/IR_strong.cat"
    print *,gettime(),"IR strong line catalogue selected"
  endif
  if (deeplinelistfile.eq."") then
    deeplinelistfile=trim(PREFIX)//"/share/alfa/IR_deep.cat"
    print *,gettime(),"IR deep line catalogue selected"
  endif
endif

if (skylinelistfile.eq."") skylinelistfile=trim(PREFIX)//"/share/alfa/sky_deep.cat"

! collapse data to 1D if requested
! apply baddata to sum only pixels with good data in
i=0 ! to count the number of pixels summed

if (collapse) then
  if (allocated(spectrum_2d)) then

    print *,gettime(),"summing rows in 2D spectrum"
    allocate(spectrum_1d(spectrumlength))
    spectrum_1d = 0.d0

    do rss_i = 1,axes(2)
      if (maxval(spectrum_2d(:,rss_i)).gt. baddata) then
        spectrum_1d = spectrum_1d + spectrum_2d(:,rss_i)
        i=i+1
      else
        bdcount=bdcount+1
      endif
    enddo

    print *,gettime(),"co-added ",i," of ",axes(2)," rows"
    if (bdcount .gt. 0) print *,gettime(),"omitted ",bdcount," rows where peak flux was less than baddata value of ",baddata
    print *

    deallocate(spectrum_2d)

  elseif (allocated(spectrum_3d)) then

    print *,gettime(),"summing pixels in 3D spectrum"
    allocate(spectrum_1d(spectrumlength))
    spectrum_1d = 0.d0

    do cube_i = 1,axes(1)
      do cube_j = 1,axes(2)
        if (maxval(spectrum_3d(cube_i,cube_j,:)) .gt. baddata) then
          spectrum_1d = spectrum_1d + spectrum_3d(cube_i,cube_j,:)
          i=i+1
        else
          bdcount=bdcount+1
        endif
      enddo
    enddo

    print *,gettime(),"co-added ",i," of ",axes(1)*axes(2)," pixels"
    if (bdcount .gt. 0) print *,gettime(),"omitted ",bdcount," pixels where peak flux was less than baddata value of ",baddata
    print *

    deallocate(spectrum_3d)

  endif
endif

!now we have all the flux values in an array and can fit the spectra

!rebin 1d spectra here

if (rebinfactor .gt. 1) then
  if (allocated(spectrum_1d)) then
    call rebinarray(wavelengths,rebinfactor)
    call rebinarray(spectrum_1d,rebinfactor)
    spectrumlength=size(wavelengths)
  endif
endif

!read in the line catalogues

print *,gettime(),"reading in line catalogues"
call readlinelist(skylinelistfile, skylines_catalogue)
call readlinelist(stronglinelistfile, stronglines_catalogue)
call readlinelist(deeplinelistfile, deeplines_catalogue)

if (allocated(spectrum_1d)) then !1d spectrum

  allocate(realspec(size(wavelengths)))
  allocate(fittedspectrum(spectrumlength))

  realspec%wavelength = wavelengths
  realspec%flux = spectrum_1d
  realspec%uncertainty = 0.d0
  redshiftguess=redshiftguess_initial
  resolutionguess=resolutionguess_initial

  fittedspectrum%wavelength=realspec%wavelength
  fittedspectrum%flux=0.d0

! bad data check

  if (maxval(realspec%flux) .le. baddata) then
    print *,gettime(),"[200] no good data in spectrum (all fluxes are less than ",baddata,")"
    call exit(200)
    stop
  endif
  messages=.true.

  threadnumber=0
  write (outputbasename,"(A)") spectrumfile(index(spectrumfile,"/",back=.true.)+1:len(trim(spectrumfile)))

  if (.not.clobber) then
    if (outputformat.eq."fits ") then
      inquire(file=trim(outputdirectory)//trim(outputbasename)//"_fit.fits", exist=file_exists)
    elseif (outputformat.eq."text ") then
      inquire(file=trim(outputdirectory)//trim(outputbasename)//"_lines", exist=file_exists)
    elseif (outputformat.eq."csv  ") then
      inquire(file=trim(outputdirectory)//trim(outputbasename)//"_lines.csv", exist=file_exists)
    elseif (outputformat.eq."latex") then
      inquire(file=trim(outputdirectory)//trim(outputbasename)//"_lines.latex", exist=file_exists)
    endif
    if (file_exists) then
      print *,gettime(),"spectrum already fitted: use --clobber to overwrite previous output"
      stop
    endif
  endif

#include "spectralfit.f90"

elseif (allocated(spectrum_2d)) then !fit 2D data

  write (filenameformat,"(A,I1,A,I1)") "I",floor(log10(real(axes(2))))+1,".",floor(log10(real(axes(2))))+1

!$OMP PARALLEL private(outputbasename,realspec,fittedspectrum,spectrumlength,continuum,nlines,spectrumchunk,linearraypos,overlap,startpos,startwlen,endpos,endwlen,skylines,skylines_section,stronglines,fittedlines,fittedlines_section,blendpeak,hbetaflux,totallines,skyspectrum,redshiftguess_overall,rss_i,threadnumber,redshiftguess,resolutionguess,file_exists,writeb1,writeb2,writep1,writep2,normalisation) shared(skylines_catalogue,stronglines_catalogue,deeplines_catalogue,axes,spectrumfile,redshiftguess_initial,resolutionguess_initial,outputdirectory)
!$OMP MASTER
  if (omp_get_num_threads().gt.1) then
    print "(X,A9,X,A,I2,A)",gettime(),"using ",omp_get_num_threads()," processors"
  endif
!$OMP END MASTER

!$OMP DO schedule(dynamic)
  do rss_i=1,axes(2) ! wavelength is on axis 1, row position is on axis 2, so this loops over rows

    threadnumber=OMP_GET_THREAD_NUM()

    write (outputbasename,"(A,"//filenameformat(1)//")") spectrumfile(index(spectrumfile,"/",back=.true.)+1:len(trim(spectrumfile)))//"_row_",rss_i

    allocate(realspec(axes(1)))
    spectrumlength=axes(1)
    realspec%wavelength = wavelengths
    realspec%flux=spectrum_2d(:,rss_i)

! rebin

    if (rebinfactor>1) then
      call rebinspectrum(realspec,rebinfactor)
      spectrumlength=size(realspec)
    endif

! initialise guesses

    redshiftguess=redshiftguess_initial
    resolutionguess=resolutionguess_initial

!check for valid data
!ultra crude at the moment

    if (outputformat.eq."fits ") then
      inquire(file=trim(outputdirectory)//trim(outputbasename)//"_fit.fits", exist=file_exists)
    elseif (outputformat.eq."text ") then
      inquire(file=trim(outputdirectory)//trim(outputbasename)//"_lines", exist=file_exists)
    elseif (outputformat.eq."csv  ") then
      inquire(file=trim(outputdirectory)//trim(outputbasename)//"_lines.csv", exist=file_exists)
    elseif (outputformat.eq."latex") then
      inquire(file=trim(outputdirectory)//trim(outputbasename)//"_lines.latex", exist=file_exists)
    endif

    if (maxval(realspec%flux) .le. baddata) then
      print "(X,A,A,I2,A,"//filenameformat(1)//",A,ES10.2)",gettime(),"(thread ",threadnumber,") : skipped row  ",rss_i,": all values below ",baddata
      deallocate(realspec)
      cycle
    elseif (file_exists .and. .not. clobber) then
      print "(X,A,A,I2,A,"//filenameformat(1)//",A)",gettime(),"(thread ",threadnumber,") : skipped row  ",rss_i,": already fitted"
      deallocate(realspec)
      cycle
    endif

    print *,gettime(),"fitting row ",rss_i

    allocate (fittedspectrum(spectrumlength))
    fittedspectrum%wavelength=realspec%wavelength
    fittedspectrum%flux=0.d0

!now do the fitting

#include "spectralfit.f90"

!deallocate arrays ready for the next pixel
    deallocate(realspec)
    deallocate(fittedspectrum)
    deallocate(continuum)
    if (allocated(skyspectrum)) deallocate(skyspectrum)

    print "(X,A,A,I2,A,"//filenameformat(1)//",A,F6.1,A,F6.0)",gettime(),"(thread ",threadnumber,") : finished row ",rss_i,". approx velocity and resolution ",c*(redshiftguess_overall-1.d0),", ",fittedlines(1)%resolution

  enddo

!$OMP END DO
!$OMP END PARALLEL

  print *,gettime(),"all processing finished"

  deallocate(spectrum_2d)

elseif (allocated(spectrum_3d)) then !fit 3D data

  print *,gettime(),"processing cube"
!$OMP PARALLEL private(outputbasename,realspec,fittedspectrum,spectrumlength,continuum,nlines,spectrumchunk,linearraypos,overlap,startpos,startwlen,endpos,endwlen,skylines,skylines_section,stronglines,fittedlines,fittedlines_section,blendpeak,hbetaflux,totallines,skyspectrum,redshiftguess_overall,cube_i,cube_j,threadnumber,redshiftguess,resolutionguess,file_exists,writeb1,writeb2,writep1,writep2,normalisation) shared(skylines_catalogue,stronglines_catalogue,deeplines_catalogue,axes,spectrumfile,redshiftguess_initial,resolutionguess_initial,outputdirectory)

!$OMP MASTER
  if (omp_get_num_threads().gt.1) then
    print "(X,A9,X,A,I2,A)",gettime(),"using ",omp_get_num_threads()," processors"
  endif
!$OMP END MASTER

  write (filenameformat(1),"(A,I1,A,I1)") "I",floor(log10(real(axes(1))))+1,".",floor(log10(real(axes(1))))+1
  write (filenameformat(2),"(A,I1,A,I1)") "I",floor(log10(real(axes(2))))+1,".",floor(log10(real(axes(2))))+1

!$OMP DO schedule(dynamic)
  do cube_i=1,axes(1)
    do cube_j=1,axes(2)

      threadnumber=OMP_GET_THREAD_NUM()

      write (outputbasename,"(A,"//filenameformat(1)//",A1,"//filenameformat(2)//")") spectrumfile(index(spectrumfile,"/",back=.true.)+1:len(trim(spectrumfile)))//"_pixel_",cube_i,"_",cube_j
      allocate(realspec(axes(3)))
      spectrumlength=axes(3)
      realspec%flux=spectrum_3d(cube_i,cube_j,:)
      realspec%wavelength=wavelengths

! rebin

      if (rebinfactor>1) then
        call rebinspectrum(realspec,rebinfactor)
        spectrumlength=size(realspec)
      endif

! initialise guesses

      redshiftguess=redshiftguess_initial
      resolutionguess=resolutionguess_initial

!check for valid data
!ultra crude at the moment

      if (outputformat.eq."fits ") then
        inquire(file=trim(outputdirectory)//trim(outputbasename)//"_fit.fits", exist=file_exists)
      elseif (outputformat.eq."text ") then
        inquire(file=trim(outputdirectory)//trim(outputbasename)//"_lines", exist=file_exists)
      elseif (outputformat.eq."csv  ") then
        inquire(file=trim(outputdirectory)//trim(outputbasename)//"_lines.csv", exist=file_exists)
      elseif (outputformat.eq."latex") then
        inquire(file=trim(outputdirectory)//trim(outputbasename)//"_lines.latex", exist=file_exists)
      endif

      if (maxval(realspec%flux) .le. baddata) then
        print "(X,A,A,I2,A,"//filenameformat(1)//",A,"//filenameformat(2)//",A,ES10.2)",gettime(),"(thread ",threadnumber,") : skipped pixel  ",cube_i,",",cube_j,": all values below ",baddata
        deallocate(realspec)
        cycle
      elseif (file_exists .and. .not. clobber) then
        print "(X,A,A,I2,A,"//filenameformat(1)//",A,"//filenameformat(2)//",A)",gettime(),"(thread ",threadnumber,") : skipped pixel  ",cube_i,",",cube_j,": already fitted"
        deallocate(realspec)
        cycle
      endif

      allocate (fittedspectrum(spectrumlength))
      fittedspectrum%wavelength=realspec%wavelength
      fittedspectrum%flux=0.d0

!now do the fitting

#include "spectralfit.f90"

!deallocate arrays ready for the next pixel

      deallocate(realspec)
      deallocate(fittedspectrum)
      deallocate(continuum)
      if (allocated(skyspectrum)) deallocate(skyspectrum)

      print "(X,A,A,I2,A,"//filenameformat(1)//",A,"//filenameformat(2)//",A,F6.1,A,F5.0)",gettime(),"(thread ",threadnumber,") : finished pixel ",cube_i,",",cube_j,". approx velocity and resolution ",c*(redshiftguess_overall-1.d0),", ",fittedlines(1)%resolution

    enddo
  enddo

!$OMP END DO
!$OMP END PARALLEL

  print *,gettime(),"all processing finished"

  deallocate(spectrum_3d)

endif

!free memory

if (allocated(axes)) deallocate(axes)
if (allocated(continuum)) deallocate(continuum)
if (allocated(deeplines_catalogue)) deallocate(deeplines_catalogue)
if (allocated(fittedlines)) deallocate(fittedlines)
if (allocated(fittedlines_section)) deallocate(fittedlines_section)
if (allocated(fittedspectrum)) deallocate(fittedspectrum)
if (allocated(originalcopy)) deallocate(originalcopy)
if (allocated(realspec)) deallocate(realspec)
if (allocated(skylines_catalogue)) deallocate(skylines_catalogue)
if (allocated(skyspectrum)) deallocate(skyspectrum)
if (allocated(spectrum_1d)) deallocate(spectrum_1d)
if (allocated(spectrum_2d)) deallocate(spectrum_2d)
if (allocated(spectrum_3d)) deallocate(spectrum_3d)
if (allocated(stronglines_catalogue)) deallocate(stronglines_catalogue)
if (allocated(wavelengths)) deallocate(wavelengths)

print *,gettime(),"all done"
print *

end program alfa
